cli/internal/daemon/daemon.go

```diff
@@ -190,26 +189,79 @@ type rpcServer interface {
 	Register(grpcServer server.GRPCServer)
 }
 
+// pidLock provides a wrapper around the pid lock file and metadata to manage
+// unlocking the pid lock.
+// TODO: consider moving into its own package?
+type pidLock struct {
+	logger   hclog.Logger
+	mu       sync.Mutex
+	lockFile lockfile.Lockfile
+	closed   bool
+}
+
+// Unlock attempts to unlock the pid lock, if it was acquired. It is idempotent, and safe to call
+// even if we didn't ever acquire the pid lock.
+func (p *pidLock) Unlock() {
+	// Since we register on the close handler before actually grabbing the pid lock,
+	// it's possible that we failed to grab it at all. It's also possible that we
+	// didn't catch any signals, and our defer ran normally at the end of runTurboServer.
+	// We handle those cases via a Mutex for synchronization, checking that p.lockFile is
+	// set to verify that we successfully acquired the lock, and the closed boolean to only
+	// attempt unlocking it once.
+	p.mu.Lock()
+	defer p.mu.Unlock()
+	if p.lockFile != "" {
+		if !p.closed {
+			p.closed = true
+			if err := p.lockFile.Unlock(); err != nil {
+				p.logger.Error(errors.Wrapf(err, "failed unlocking pid file at %v", p.lockFile).Error())
+			}
+		}
+	}
+}
+
+func acquirePidLock(pidPath fs.AbsolutePath, signalWatcher *signals.Watcher, logger hclog.Logger) (*pidLock, error) {
+	if err := pidPath.EnsureDir(); err != nil {
+		return nil, err
+	}
+	p := &pidLock{
+		logger: logger,
+	}
+	p.mu.Lock()
+	defer p.mu.Unlock()
+	// If we catch a signal, unlock the pid file. This also wires up
+	// the pidfile for unlock on program exit.
+	signalWatcher.AddOnClose(func() {
+		p.Unlock()
```

Comment:
In case I don't get to implementing this, leaving some notes here. 

I did some experimentation with sending signals at different points in the execution of this function. It's also important to note that we are *not* listening to `SIGKILL`. Since registering a signal handler alters the default behavior of the runtime, we can block indefinitely on getting a `Ctrl+C` / `SIGINT`. This means that rather than needing to register watchers for signals from `runTurboServer`, we can instead add the signal watcher's `Done()` channel to the `select` at the bottom.

This definitely simplifies things, as our `defer lock.Unlock()` will always run. I also think it can be the only wait runs: no matter which method of triggering occurs (signal, request handler panic, server goroutine panic), the end result is breaking out the select at the bottom of `runTurboServer`. With that in mind, I *think* we don't noeed the `pidLock` abstraction. Future work that might apply would be wiring up a `context.NotifyContext` and dropping the `signals.Watcher` abstraction.